app/javascript/packs/leafletMap.js

```diff
@@ -114,38 +127,54 @@ async function getView() {
-        permanent: true,
-        direction: "center",
-        className: "indoor-label",
-    });
+function setupGeoJsonFeature(feature, layer) {
+    const level = parseInt(feature.properties.level_name);
+    if (Number.isInteger(level)) {
+        if (!floors[level]) {
+            floors[level] = { rooms: L.layerGroup(), labels: L.layerGroup() };
+            const layer = L.layerGroup([floors[level].rooms, floors[level].labels]);
+            layerControl.addBaseLayer(layer, feature.properties.level_name);
+            // We add the current floor to the map here so that the map and layer control reference the same object 
+            // now the layer controll will select the correct check box automatically
+            if (level == currentFloor) {
+                layer.addTo(map);
+            }
+        }
+
+        floors[level].rooms.addLayer(layer);
+
+        // find a position, create a invisible marker and bind a tooltip to it
+        // we do not bind the tooltip to the room because you can only change the position of a tooltip by defining an offset 
+        // to the referenced object (you cannot give it a coordinate), we would need to determine an offset for each room
+        const markerPosition = polylabel(feature.geometry.coordinates, 0.000001);
+
+        const label = L.circleMarker({ lat: markerPosition[1], lng: markerPosition[0] }, { opacity: 0, radius: 0 }).bindTooltip(
+            feature.properties.name,
+            {
+                permanent: true,
+                direction: 'center',
+                className: 'indoor-label',
+            }
+        )
+
+        floors[level].labels.addLayer(label);
+        label.closeTooltip();
+    }
 }
 
 function loadGeoJsonFile(filename) {
-    fetch(filename)
-        .then((response) => response.json())
-        .then((geojsonFeatureCollection) => {
-            // Manually add indoor labels to map
-            const rooms = L.geoJSON(geojsonFeatureCollection, {
-                onEachFeature: addIndoorLabel,
-            }).addTo(map);
-            rooms.eachLayer((layer) => {
-                layer.getTooltip().setLatLng(layer.getBounds().getCenter());
-            });
-            recalculateTooltipVisibility();
-        });
+    return fetch(filename)
+        .then(response => response.json())
+        .then(geojsonFeatureCollection => {
+            L.geoJSON(geojsonFeatureCollection, { onEachFeature: setupGeoJsonFeature, filter: (f => { return f.geometry.type != "Point" }) });
+        })
 }
 
 function recalculateTooltipVisibility() {
-    const zoomLevel = map.getZoom();
-    map.eachLayer((layer) => {
-        if (layer.getTooltip()) {
-            if (zoomLevel == 19 /* nearest zoom */) {
-                layer.openTooltip(layer.getBounds().getCenter());
-            } else {
-                layer.closeTooltip();
-            }
-        }
-    });
+    const zoomLevel = map.getZoom()
+    if (zoomLevel >= 19) {
```

Comment:
In my opinion we should only start showing the labels from level `20` onwards as they are just too small at level `19`. This is just my opinion though, maybe this is a question for @anonym-HPI? Currently, these are the layers on level 19 (which I suggest to not display): 
![Bildschirmfoto von 2022-01-15 19-08-03](https://user-images.githubusercontent.com/68013019/149632960-ebd5b4e4-92a2-441a-bb27-5429774817e7.png)

and these are the layers on level 20 (which I suggest to display): 
![Bildschirmfoto von 2022-01-15 19-08-09](https://user-images.githubusercontent.com/68013019/149632976-456d4825-1a8d-4f64-879c-d237be3a39d7.png)
What do you think?

Reply:
agree!